README.md
readGenalex
===========

An R package for reading data files in GenAlEx format, as exported from Excel
as a delimited text file, into an annotated `data.frame` of class `'genalex'`,
and manipulate it in that form.  Several functions are provided for accessing,
manipulating, and printing this data.  GenAlEx and its documentation are
available at <http://biology-assets.anu.edu.au/GenAlEx>.

The most recent release 1.0 (now on CRAN) introduces the S3 class `'genalex'`,
and implements the functionality of this package by defining an S3 class
`'genalex'` that is shared with `data.frame`.



## Release version 1.0.9000

The development version is hosted on Github and can always
be installed via:

```R
> install.packages("devtools")
> devtools::install_github("douglasgscofield/readGenalex")
```


## Using the package

`readGenalex` 1.0 can be installed through CRAN:

```R
> install.packages("readGenalex")
```

Class `genalex` is an annotated data frame with attributes containing
additional information specified by the user:

~~~~
> library(readGenalex)
> data(Qagr_adult_genotypes)
> head(Qagr_adult_genotypes)
  AdultMomFamily     Site 0c11 0c11.2 0c19 0c19.2 Oe09 Oe09.2 0i01 0i01.2 0m05
1           2201 Sedgwick  215    215  226    244  190    192  204    206  210
2           2202 Sedgwick  213    217  238    238  190    192  204    204  210
3           2203 Sedgwick  213    215  234    240  190    192  196    206  216
4           2204 Sedgwick  213    213  234    234  186    192  196    204  210
5           2205 Sedgwick  213    217  222    238  190    194  204    204  210
6           2206 Sedgwick  213    213  226    240  194    194  196    204  202
  0m05.2 0m07 0m07.2 1c06 1c06.2 1c08 1c08.2 1f02 1f02.2 1g13 1g13.2
1    210  199    199  238    242  273    273  190    190  189    191
2    222  199    205  244    244  273    273  190    190  185    187
3    220  199    199  234    234  273    273  190    190  189    189
4    212  199    201  234    236  269    273  190    190  187    191
5    216  201    205  236    242  273    273  180    190  185    187
6    206  201    203  240    242  271    273  190    190  185    185
> class(Qagr_adult_genotypes)
[1] "genalex"    "data.frame"
> attributes(Qagr_adult_genotypes)
$names
 [1] "AdultMomFamily" "Site"           "0c11"           "0c11.2"
 [5] "0c19"           "0c19.2"         "Oe09"           "Oe09.2"
 [9] "0i01"           "0i01.2"         "0m05"           "0m05.2"
[13] "0m07"           "0m07.2"         "1c06"           "1c06.2"
[17] "1c08"           "1c08.2"         "1f02"           "1f02.2"
[21] "1g13"           "1g13.2"

$row.names
  [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
 [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
 [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
 [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
 [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
 [91]  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
[109] 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
[127] 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
[145] 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
[163] 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180
[181] 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198
[199] 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216
[217] 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234
[235] 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251 252
[253] 253 254 255 256 257 258 259 260 261 262

$class
[1] "genalex"    "data.frame"

$n.loci
[1] 10

$ploidy
[1] 2

$n.samples
[1] 262

$n.pops
[1] 1

...
~~~~

`readGenalex` reads GenAlEx-format data from a text file, and
`readGenalexExcel` reads data from a specified worksheet of an Excel workbook
of `.xls` or `.xlsx` format using the `XLConnect` package.  Both functions read
only the number of samples specified by the GenAlEX header, and only treat as
genotypes the number of genotype columns implied by the GenAlEx header in
concert with the stated ploidy level.

Because GenAlEx is an Excel plugin, `readGenalex` attempts to deal with various
issues that may arise when exporting text files from Excel.  `readGenalex` also
tries to ignore extra tab characters that tools such as Excel can insert when
exporting tab-delimited text, otherwise these could imply both additional
columns and additional rows.  Hopefully the latter is avoided by only reading
the number of samples specified by the header.

If there are additional **named** columns to the right of the genotypes, these
are read and stored in a data frame attached to the attribute `extra.columns`.
The first column of the `extra.columns` data frame is the sample name (leftmost
column from the genotypes, e.g., the `id` column from the above example).  It
attempts to ignore additional unnamed columns scattered amongst the named extra
columns.

There are also corresponding `writeGenalex` and `writeGenalexExcel` functions.

Functions:

Function | Description
-------- | -----------
`readGenalex()` | Read GenAlEx-format data from a text file
`readGenalexExcel()` | Read GenAlEx-format data from a worksheet of an Excel workbook
`genalex()` | Create a class `'genalex'` object from constituent data
`as.genalex()` | Generic function which converts a pre-1.0-style `readGenalex` data frame to class `'genalex'`, or converts a suitably-formatted data frame to class `'genalex'`.  Optionally it can determine structure implied by the data and update attributes to reflect that structure.
`writeGenalex()` | Write a GenAlEx-format text file
`writeGenalexExcel()` | Write a GenAlEx-format worksheet to an Excel workbook
`summary()` | Prints a summary of the data set, a summary of the genotype data frame, and a summary of the extra columns, if any
`is.genalex()` | Checks whether the object is class `'genalex'`, optionally does a deeper check to determine whether the structure described in the attributes matches the structure implied by the data
`as.data.frame()` | Method to convert class `'genalex'` to class `'data.frame'`, optionally all class `'genalex'`-specific attributes are removed
`reducePloidy()` | Reduce the ploidy to 1 by selecting the first allele of each locus
`dropLocus()` | Drop named loci
`getPopulation()` | Return genotypes of specific populations in object of class `'genalex'` 
`printGenotype()` | Print genotypes of specific rows
`reorderLoci()` | Reorder loci into a given order
`computeLocusColumns()` | Return a vector of column numbers for specified loci
`replaceLocus()` | Replace genotypes of specified locus
`getLocus()` | Return genotypes of specified locus, optionally only for specific populations
`addLocus()` | Add genotypes to an object of class `'genalex'`
`extra()` | Get or set `extra.columns` attribute
`ploidy()` | Get `ploidy` attribute
`cbind()` | Merge loci and extra data columns from two or more class `'genalex'` objects
`rbind()` | Merge samples from two or more class `'genalex'` objects
`checkNullAlleles()` | Compare genotypes against a set of reference genotypes to check for potential null (nonamplifying) alleles
`writeGenepop()`| Write class `'genalex'` object in [Genepop](http://kimura.univ-montp2.fr/~rousset) format
`as.genetics()` | Convert class `'genalex'` object to a data frame with genotypes encoded using class `genotype` (unphased) or class `haplotype` (phased) from package `genetics`
`splitGenotypes()` | Split genotypes encoded as `101/107` into separate columns of a data frame, suitable for further use with `genalex()`
`joinGenotypes()` | Join class `'genalex'` (or other class) genotypes into a single column
`as.loci.genalex` | Extends the `as.loci` generic from the `pegas` package to convert class `'genalex'` to class `'loci'`
`as.genalex.loci` | Converts an object of class `'loci'` from the `pegas` package to class `'genalex'`


## Datasets

The package also provides two data sets as class `'genalex'` objects that can be loaded with `data`:

* `Qagr_adult_genotypes`: Coast live oak (*Quercus agrifolia*) adult microsatellite genotypes, holding 10-locus diploid microsatellite genotypes of 262 adult coast live oak trees from Sedgwick Reserve, Santa Barbara County, California, USA
* `Qagr_pericarp_genotypes`: Coast live oak (*Quercus agrifolia*) pericarp microsatellite genotypes, holding 10-locus diploid microsatellite genotypes of 568 pericarps (outer seed coats) from coast live oak acorns collected from 17 acorn woodpecker (*Melanerpes formicivorus*) granaries at Sedgwick Reserve, Santa Barbara County, California, USA

These data sets are also available at the Dryad Data Repository, <http://datadryad.org/resource/doi:10.5061/dryad.40kq7>.

Several papers have been published with these data.  If using them, please cite the original paper as well as the data:

Scofield DG, Smouse PE, Karubian J, Sork VL (2012) Use of alpha, beta, and gamma diversity measures to characterize seed dispersal by animals. *The American Naturalist* 180(6): 719-732.  <http://dx.doi.org/10.1086/668202>

Scofield DG, Smouse PE, Karubian J, Sork VL (2012) Data from: Use of alpha, beta, and gamma diversity measures to characterize seed dispersal by animals.  Dryad Digital Repository. <http://dx.doi.org/10.5061/dryad.40kq7>


NEWS
readGenalex 1.0.9000
====================

-   Documentation tweaks

readGenalex 1.0
===============

-   Creates and manipulates data frames using the new class 'genalex'.
    The "genetic.data.format" attribute present in earlier versions of
    this package is removed.
-   Extensive documentation updates reflecting use of class 'genalex'
-   Several functions have been deprecated and replaced with generics
    recognising class 'genalex': printGenalexGenotype is replaced by
    printGenotype, computeGenalexColumns is replaced by getLocusColumns,
    reorderGenalexLoci is replaced by reorderLoci, getGenalexLocus is
    replaced by getLocus, putGenalexLocus is replaced by replaceLocus,
    dropGenalexLoci is replaced by dropLocus, and reduceGenalexPloidy is
    replaced by reducePloidy
-   Added getPopulation to return genotypes for specific populations in
    a class 'genalex' object
-   Extended is.genalex to have shallow (default) or deep (force = TRUE)
    checking of class 'genalex' consistency. With deep checking,
    attributes are inferred from the structure of the data frame and
    compared to the actual attributes. Checking of string attributes
    (dataset.title, data.file.name) and extra data columns may skipped
    with options.
-   New as.genalex generic with as.genalex.genalex and
    as.genalex.data.frame. The latter will convert a pre-class 'genalex'
    data frame to class 'genalex', and is used wherever a data frame is
    returned so conversion from old to new format happens automatically.
    It will also convert a suitably-formatted data frame to class
    'genalex' after determining suitable attributes
-   There is a summary method for class 'genalex' which prints a few
    lines describing the data set, followed by a summary of the data
    frame contents and a summary of the extra columns, if present
-   New as.data.frame method to convert class 'genalex' to class
    'data.frame', optionally clearing all class 'genalex'-specific
    attributes
-   New as.genetics method to convert class 'genalex' to a data frame
    with genotypes for each locus encoded using class 'genotype' or
    class 'haplotype' from package genetics
-   New splitGenotypes function to split genotype-encoded genotypes into
    separate columns of a data frame
-   New joinGenotypes function to join class 'genalex'-style allele
    columns into a single column per locus
-   New as.loci.genalex method for the as.loci generic from package
    pegas to convert a class 'genalex' object to class 'loci', and
    as.genalex.loci for the reverse conversion
-   New genalex function to create a class 'genalex' object from
    constituent data
-   New rbind method to combine samples from multiple class 'genalex'
    objects
-   New cbind method to combine loci and extra data columns from
    multiple class 'genalex' objects
-   New addLocus method to directly add genotype data to the right side
    of an existing class 'genalex' object
-   New extra and extra<- methods to get and set the extra columns of a
    class 'genalex' object. The latter coerces its value object to a
    data frame, if it is not already, does a bit of error checking, and
    sets the row names to be the sample names of the object.
-   New ploidy method to get the ploidy of an object of class 'genalex'
-   readGenalex now has ... in its argument list, for additional
    arguments to be passed to scan() while reading data
-   Added na.strings argument to readGenalex, with defaults that reflect
    GenAlEx's own expectations for missing data
-   Added na and na.character arguments to writeGenalex, with values to
    use for NA in genotype/numeric/logical columns and character
    columns, respectively
-   Added readGenalexExcel and writeGenalexExcel which use the XLConnect
    package to read and write to Excel files of both .xls and .xlsx
    formats. These functions verify the consistency of their annotations
    and data prior to writing; this check can be removed on option.
-   Added writeGenepop to write data in a class 'genalex' object to a
    Genepop-format text file
-   Clarified typing of sample, population, genotype, and extra columns.
    Sample and population columns are stored as character (population
    was previously a factor), genotype columns are stored as numeric,
    and extra columns are stored as their natural type as determined by
    type.convert(..., as.is=TRUE), so that character columns are not
    converted to factors.
-   As a result of the clarified typing, writeGenalex(..., quote=TRUE)
    will only quote names in the GenAlEx header, in the first two
    columns, and in any character-valued extra columns
-   There is now more thorough checking for duplicate sample names,
    which should always produce an error
-   There are also checks for duplication among the locus names and the
    sample and population columns
-   The locus.columns attribute now has names which match the
    corresponding loci
-   Added tests for all functions and found a few bugs along the way
-   Rolled checkNullAlleles from my popgen repository into this package
-   Extended checkNullAlleles to return a matrix of potential null
    alleles, with a couple of indexing modes
-   Added quiet= option to checkNullAlleles, and pass through some
    arguments to printGenotype when quiet=FALSE
-   Added Qagr_adult_genotypesand Qagr_pericarp_genotypes data sets and
    modified examples to use them, and removed the example_genotypes
    data set
-   Created inst/extdata to hold data in non-R formats (simple tables),
    which of course includes GenAlEx-format data
-   DESCRIPTION now has an explicit collation order and bug report URL
-   Added 'data' target to Makefile to build specified RData files

readGenalex 0.4.1.9000 (development version)
============================================

-   Moved documentation from separate *.Rd files to Roxygen2 comments in
    R source using the really helpful Rd2roxygen package.

R/readGenalex-pegas.R
#' @include readGenalex.R
#' @include readGenalex-readwrite.R
#' @include readGenalex-genetics.R
# for collation order
NULL

#' Convert class \code{genalex} object to data frame of class \code{'loci'} from package \code{'pegas'}
#'
#' Converts an object of class \code{genalex} to a data frame of class
#' \code{'loci'} from the
#' \href{http://cran.r-project.org/web/packages/pegas/index.html}{pegas}
#' package.  This data frame is similar to one of class \code{genalex},
#' in that it mixes genetic and other data in the same data frame, but
#' its conversion of multiple allele columns to single genotype columns
#' is similar to the result of the \code{\link{as.genetics}} function
#' of this package.
#'
#' Like class \code{genalex}, class \code{'loci'} can encode genotypes of
#' any ploidy.  Once a class \code{genalex} object is converted to class
#' \code{'loci'}, it may be further converted to other data structures for
#' analysis with
#' \href{http://cran.r-project.org/web/packages/pegas/index.html}{pegas}
#' and
#' \href{http://cran.r-project.org/web/packages/adegenet/index.html}{adegenet}.
#'
#' The specific changes that occur to an object of
#' class \code{genalex} for it to become an object of class \code{'loci'}:
#' \itemize{
#'    \item Row names are set from sample names
#'    \item The first column of sample names is removed
#'    \item The population column is renamed \code{"population"}, with
#'          its original name retained in the \code{"pop.title"} attribute.
#'          It is also encoded as a factor.
#'    \item The individual alleles of a locus are merged into a single
#'          column, with alleles separated by \code{"/"}
#'          (\code{phased = FALSE}) or \code{"|"} (\code{phased = TRUE}).
#'          These columns are encoded as factors.
#'    \item The \code{"locus.columns"} attribute is updated to reflect that
#'          all alleles at a locus are now joined into a single column
#'    \item A new attribute \code{"locicol"} required by class \code{'loci'}
#'          is added, with a value identical to the \code{"locus.columns"}
#'          attribute
#'    \item The \code{class} is changed from \code{c(genalex, 'data.frame')}
#'          to \code{c('loci', 'data.frame')}
#' }
#'
#' Because of the removal of the sample name column and the additional allele
#' columns, the number of columns will be reduced by 1 plus the number of loci.
#' For further details of the structure of class \code{\link[pegas]{loci}}, see
#' \url{http://ape-package.ird.fr/pegas/DefinitionDataClassesPegas.pdf}.
#'
#' Because class \code{'loci'} can readily encode additional columns, the
#' extra columns of a class \code{genalex} object can be bound with
#' \code{cbind} as additional columns.
#'
#' This is a specialised wrapper around the function
#' \code{\link[pegas]{as.loci.data.frame}} from the
#' \href{http://cran.r-project.org/web/packages/pegas/index.html}{pegas}
#' package.
#'
#' @param x       Annotated data frame of class \code{genalex}
#'
#' @param phased  Still some details to work out.  Default \code{FALSE}.  If
#' \code{FALSE}, assumes alleles in \code{x} are unphased so that a genotype
#' of \code{101/107} is identical to a genotype of \code{107/101}.  This
#' results in the use of \code{"/"} as the allele separator.  If \code{TRUE},
#' uses \code{"|"} as the allele separator, and assumes alleles are phased so
#' that a genotype of \code{101|107} is different from a genotype of
#' \code{107/101}.
#'
#' @param check.annotation  If \code{TRUE}, the annotations for the dataset
#' are checked using \code{is.genalex(x, force = TRUE, skip.strings = TRUE)}
#' prior to conversion.  If that returns \code{FALSE}, nothing is converted
#' and an error is generated.
#'
#' @param \dots   Additional arguments, currently ignored
#'
#' @return \code{x} as an object of class \code{'loci'}, which is a data
#' frame with the genotype of each locus encoded as factors.  Additional
#' changes apply, see Details.
#'
#' @author Douglas G. Scofield
#'
#' @seealso  \code{\link[pegas]{as.loci}}, \code{\link{joinGenotypes}}
#'
#' @examples
#'
#' suppressPackageStartupMessages(require(pegas))
#' data(Qagr_pericarp_genotypes)
#' dd <- as.genalex(head(Qagr_pericarp_genotypes, 15), force = TRUE)
#' as.loci(dd)
#' str(as.loci(dd, phased = TRUE))
#'
#' @export
#'
as.loci.genalex <- function(x, phased = FALSE, check.annotation = TRUE, ...)
{
    if (! requireNamespace("pegas", quietly = TRUE))
        stop("Please install package 'pegas' to use this function")
    sep <- if (phased) "|" else "/"
    x.name <- deparse(substitute(x.name))
    if (check.annotation && ! is.genalex(x, force = TRUE, skip.strings = TRUE))
        stop(x.name, " class 'genalex' annotations are inconsistent, not converting")
    x.attr <- attributes(x)
    x <- joinGenotypes(x, sep = sep)  # x now data.frame
    # remove sample names column
    rownames(x) <- x[, 1]
    x[, 1] <- NULL
    attr(x, "locus.columns") <- attr(x, "locus.columns") - 1
    # make the genotypes factors
    for (col in attr(x, "locus.columns"))
        x[, col] <- as.factor(x[, col])
    # convert to class 'loci'
    x <- pegas::as.loci(x, allele.sep = "/|", col.pop = 1,
                        col.loci = attr(x, "locus.columns"))
    attr(x, "data.file.name") <- paste0("as.loci(", x.attr$data.file.name, ")")
    x
}



# See R/readGenalex.R for as.genalex.loci



#' Join genotypes encoded as separate alleles into single genotypes like class \code{'genotype'} or class \code{'loci'}
#'
#' @param  x   An object of class \code{genalex}, a data frame, or an
#' object that can be coerced to a data frame which contains genotypes
#' with alleles in separate columns, to be joined into character
#' genotypes of the form \code{"101/107"}, where the \code{"/"} is
#' determined by \code{sep}.
#'
#' @param  sep     Separator between alleles
#'
#' @param  loci    Columns or column names indicating loci to join.  When
#' called for an object of class \code{genalex}, all loci are joined.
#'
#' @param  ploidy  Ploidy of the indicated loci.  When called for an
#' object of class \code{genalex}, ploidy is determined from attributes.
#'
#' @param \dots   Additional arguments, currently ignored
#'
#' @return A data frame with allele columns replaced by genotype columns
#'
#' @export
#'
#' @name joinGenotypes
#'
NULL

joinGenotypes <- function(x, ...) UseMethod("joinGenotypes")

#' @rdname joinGenotypes
#'
#' @export
#'
joinGenotypes.genalex <- function(x, sep = "/", ...)
{
    x.attr <- attributes(x)
    x <- as.data.frame(x)
    x <- joinGenotypes.default(x, loci = x.attr$locus.columns,
                               ploidy = x.attr$ploidy, sep = sep, ...)
    x.attr$names <- c(x.attr$names[1:2], x.attr$locus.names)
    x.attr$locus.columns <- setNames(.calculateLocusColumns(x.attr$n.loci, 1),
                                     x.attr$locus.names)
    x.attr$data.file.name <- paste0("joinGenotypes(", x.attr$data.file.name, ")")
    x.attr$class <- 'data.frame'
    attributes(x) <- x.attr
    x
}



#' @rdname joinGenotypes
#'
#' @export
#'
joinGenotypes.default <- function(x, loci, ploidy, sep = "/", ...)
{
    if (! is.data.frame(x))
        x <- as.data.frame(x)
    if (missing(loci) || missing(ploidy))
        stop("both 'loci' and 'ploidy' must be supplied")
    if (is.character(loci))
        loci <- match(loci, names(x))
    locus.names <- names(x)[loci]
    drop.cols <- c()
    for (col in loci) {
        allele.cols <- seq(col, (col + ploidy - 1), 1)
        m <- x[, allele.cols, drop = FALSE]
        m <- apply(m, 1, paste, collapse = sep)
        drop.cols <- c(drop.cols, allele.cols[-1])
        x[, allele.cols[1]] <- m
    }
    x[, -drop.cols]
}

